### DISCUSSION TITLE
Unable to make it work with Cloudflare Agro Tunnel, hoping someone can help.

### DISCUSSION THREAD

**Discussion by apildk** [llamafilm/tesla-http-proxy-addon#115] [Q&A]
Hi! 
First of all thanks for reading, i hope someone are able to see easily where i have gone wrong in trying to make this work.

This all came about as suddenly my tesla integration stopped being able to change things such as starting charging etc, but came up with an error about "Failed to call service switch/turn_on. unknown error" when i try.
Polling data still works.

So after trying everything and digging through google about what the issue is, i came to that i think my issues would be solved with this http proxy addon functioning. (please correct me if i am wrong)

My domain setup is a domain with cloudflare manageing the dns, and using their tunnel to access my HA.
I have read through many topics and i found this guide: https://github.com/llamafilm/tesla-http-proxy-addon/wiki/Cloudflare-Argo-Tunnel
which i based my setup and configuration on.

I am unable to access the public-key.pem (getting a Error code 502) when i try the url for this, so this must be where my problem lays.

NGINX SSL Proxy:

My Nginx ssl proxy seems to be ok, as log is empty and if i restart it i dont get any errors.
The config of that is :
domain: homeassistant-internal.
hsts: max-age=31536000; includeSubDomains
certfile: fullchain.pem
keyfile: privkey.pem
cloudflare: true
customize:
  active: true
  default: nginx_proxy_default*.conf
  servers: nginx_proxy/*.conf
  

Cloudflare:

My config is:
external_hostname: ha.
additional_hosts:
  - hostname: tesla.
    service: https://homeassistant-internal.
    
in the log it is spammed with this sorta log:
2024-06-12T19:02:05Z ERR  error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=3 event=1 ingressRule=1 originService=https://homeassistant-internal.
2024-06-12T19:02:05Z ERR Request failed error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=3 dest=https://tesla./.well-known/appspecific/com.tesla.3p.public-key.pem event=0 ip=198.41.200.13 type=http

Let's encrypt

This seems to be ok, as after setting up the correct priviliages for the api token, then it managed to get the certificates, and said all was ok, and stopped it self.
Config:
domains:
  - ha.
  - tesla.
  - homeassistant-internal.
email: my@mail.com
keyfile: privkey.pem
certfile: fullchain.pem
challenge: dns
dns:
  provider: dns-cloudflare
  cloudflare_api_token: averylongapitokenfromcloudflare

Tesla HTTP Proxy:

Config:
client_id: myclintid
client_secret: myclintsecrect
domain: tesla.domain.com
debug: false
regenerate_auth: true
region: Europe, Middle East, Africa

Here i am getting this log:
HTTP/2 502 
date: Wed, 12 Jun 2024 19:02:05 GMT
curl: (22) The requested URL returned error: 502
Warning: Problem : HTTP error. Will retry in 256 seconds. 6 retries left.
content-type: text/plain; charset=UTF-8
content-length: 15
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v4?s=somelongtokencodemaybe"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
x-frame-options: SAMEORIGIN
referrer-policy: same-origin
cache-control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0
expires: Thu, 01 Jan 1970 00:00:01 GMT
server: cloudflare
cf-ray: mayebserectidunno
alt-svc: h3=":443"; ma=86400

I have worked on this for days now, and am unable to figure out where my problem lays.
So i would be very grateful if anyone is able to help me out.

Thank you for reading and your responds.

Best wishes!

**Comment by baylanger** [comment]
Can you please use the config as specified in the Wiki?

For example, in many places you are using "domain.com" instead of "ha.domain.com". Call it "home-assistant" or "hass" or anything but don't leave it as "domain.com". Perhaps the issue is somewhere else. After doing that, if you still need help please post the FQDN you use internally and externally.

**Reply by apildk** [reply]
Hi Baylanger, thanks for your reply.

I tried now to change the main domain of access to my homeassistant to ha."myfqdn.com", but the issue still persist.

I do not understand why you need my FQDN to futher help me ?

These are the logs 

From Cloudflare addon:

2024-06-16T18:23:31Z ERR  error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=2 event=1 ingressRule=1 originService=https://homeassistant-internal.myfqdn.com
2024-06-16T18:23:31Z ERR Request failed error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=2 dest=https://tesla.myfqdn.com/.well-known/appspecific/com.tesla.3p.public-key.pem event=0 ip=198.41.200.73 type=http

From Tesla http proxy:

curl: (22) The requested URL returned error: 502
Warning: Problem : HTTP error. Will retry in 256 seconds. 6 retries left.
HTTP/2 502 
date: Sun, 16 Jun 2024 18:23:31 GMT
content-type: text/plain; charset=UTF-8
content-length: 15
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v4?s=SXen8lKKn62GFJ4PJ0entZnnnOq40POu8G1Tfi1junlobjZ%2FCu3cv%2BOPWgfk0%2Fk9tdqWaXK6OpcQKuxlG%2Fm3ErHo2szLZvAw4DcmKOoUqao7qQkwvPpKZprfdZSHkmlS"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
x-frame-options: SAMEORIGIN
referrer-policy: same-origin
cache-control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0
expires: Thu, 01 Jan 1970 00:00:01 GMT
server: cloudflare
cf-ray: 894cd4dfeb3b92b5-CPH
alt-svc: h3=":443"; ma=86400

Thanks agian! :)

-Apil

**Comment by gandhiarpit** [comment]
I have the exact same issue and can't figure out what is going on. Have followed the wifi steps for argo tunnel to the T 

Cloudflare log
```
2024-06-17T09:28:19Z ERR  error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=2 event=1 ingressRule=1 originService=https://tesla-internal.mydomain.net
2024-06-17T09:28:19Z ERR Request failed error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=2 dest=https://tesla.mydomain.net/.well-known/appspecific/com.tesla.3p.public-key.pem event=0 ip=198.41.200.53 type=http
```
Tesla proxy log
```
HTTP/2 502 
date: Mon, 17 Jun 2024 09:13:19 GMT
content-type: text/plain; charset=UTF-8
content-length: 15
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v4?s=4t2dCETsrnKquKw2nl3FlqEUwOHgEOKWpOCHdt9rJ4ZyJ9Rj3CzIPE1KQeET03kE%2FpGHu5ydVv7r5Bt%2BI6F%2FR%2BXDJlozeRPdj6J8xILZYMrU%2FCAota0%2BWo2M0fh67cXZ3vCTdJyjkF7xqsHw%3D%3D"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
x-frame-options: SAMEORIGIN
referrer-policy: same-origin
cache-control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0
expires: Thu, 01 Jan 1970 00:00:01 GMT
server: cloudflare
cf-ray: 8951ec465c7ca80d-SYD
alt-svc: h3=":443"; ma=86400
```

**Comment by gandhiarpit** [comment]
Finally got it sorted. Not sure if this is what fixed it or not as tried too many things. But one of the main thing I did is copying tesla public key.pem & selfsigned.pem files to below folder as well.

This are created under `/share/tesla`
Copy them to `/homeassistant/www/tesla -> both *.pem files.`

![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/7076783/d71324b0-105e-4692-a9db-080c88b67c5f)

I am running HAOS VM on proxmox. I tried to use my existing Cloudflared lxc running on Proxmox, but that didn't work somehow. Had to spin up another instance of HA as something on my main HA kept failing ssl proxy which brought down my Proxmox host. I am not good at network side of things so didn't know what happened. Now I have it working, on another HA instance. So just created a new tunnel just for HA and finally after 3 days of figuring this out, finally got it working.

Thanks.

**Reply by apildk** [reply]
Hi! and thanks for sharing!

Did you create the www/ tesla folder your self in homeassistant/ or was it already there ?
(I do not have this folder currently)

Also where did you find the two .pem files and copy them from ?

Best regards
-Apil

**Reply by gandhiarpit** [reply]
Tesla proxy creates them under `/share/tesla`. I think it is mentioned somewhere in documentation.

If www is not there, create it. No harm trying it. I don't recall if I created or it was there. 

![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/7076783/f0006477-f299-45c4-8d67-e29c37006015)

**Reply by apildk** [reply]
Unfortunately i dont think this is the issue.
I copied the two files from /share/tesla and created a folder named tesla under /www and placed the files in it.
then restart of HA and still same issue :(
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/b76569b9-e50e-473c-a8ce-6e1d2140be82)

**Reply by gandhiarpit** [reply]
One other thing which I had to do was changing SSL Edge certificates to flexible. I had strict set and that didn't work(considering lots of changes, one of them)
`Check on the Cloudflare dashboard that in the SSL/TLS Edge Certificates there's an active universal certificate for both example.com and *.example.com. Make sure that in the SSL/TLS Overview page the encryption mode is set to Flexible`

I had nginx running on my unraid server which somehow broke by above change so had to move some of my services to tunnel. But it could be just my setup.

Sorry. That's all I can help with.

**Reply by apildk** [reply]
Thanks again for taking the time to try and help me out and replying.
Though in my setup i do have the flexiable option and edge certificates on both.
So that is not my problem :(
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/aa1a4630-2b3a-4041-b5d7-17c06e0520bd)

**Comment by apildk** [comment]
Hi again.

I am still hoping someone would help me out here.

I am trying to figure out where it goes wrong, and as far as i can tell the problem is between cloudflare and the tesla http proxy.
Cloudflare sees the call to the the origin service at the "https://homeassistant-internal.mydomain.com" but unable to reach it
Which is setup in cloudflare to point to the local ip address of my home assistant 192.168.100.3 as it says in the guide.
(i run everything on a homeasistant rasberry pi, with eveything as addons in HA)

I do not understand how it is suppose to get to the tesla http proxy addon.
I see that when the tesla http proxy addon is starting up then it says it is lisning on:
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.3:8099

And i do see in my config that i have:
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - 172.30.33.0/24
I dont know if this has anything to do with it.

I am able to reach my homeassistant login page when i try the "homeassistant-internal" url, so that should all be good.

Can someone help me and maybe explain to me how it is suppose to reach the tesla http proxy addon with this config ?
(https://github.com/llamafilm/tesla-http-proxy-addon/wiki/Cloudflare-Argo-Tunnel)

Best regards

**Reply by gandhiarpit** [reply]
My assumption is that you don't need below x_forwarded in configuration.yaml as you are using CF. I believe that setting was for duckdns.

`use_x_forwarded_for: true`

**Reply by apildk** [reply]
removing the "use_x_forwarded_for: true" gave a lot of errors, and HA went into recovery mode, with these logs:
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/3edd9109-3315-40b0-9cc4-bb18ebd39a3f)

**Comment by apildk** [comment]
@baylanger I have updated the thread with my domain, as i am desperate to get this working.
(my domain is apil.dk)
ha.
tesla.
homeassistant-internal.
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/fe4e2eb8-c998-4899-bcc6-68d4163dd5d9)
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/ed3f6f71-1135-4d09-909a-731d1ce692cc)

Cloudflare Logs:
2024-06-21T14:20:26Z ERR  error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=1 event=1 ingressRule=1 originService=https://homeassistant-internal.
2024-06-21T14:20:26Z ERR Request failed error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: EOF" connIndex=1 dest=https://tesla./.well-known/appspecific/com.tesla.3p.public-key.pem event=0 ip=198.41.200.23 type=http

Tesla http proxy logs:
HTTP/2 502 
date: Fri, 21 Jun 2024 14:20:26 GMT
content-type: text/plain; charset=UTF-8
content-length: 15
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v4?s=NpSv1k0REE3CyCqFoCWYY5DzsGQOhWHv1%2FK%2FeEa4RVcrIJSbEV7waw47NSzurd0JLAIOsp6cbHI2RiTYiPzw3waxgUC5T8bK1Hf3KAltPpNzyH99y"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
x-frame-options: SAMEORIGIN
referrer-policy: same-origin
cache-control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0, pre-check=0
expires: Thu, 01 Jan 1970 00:00:01 GMT
server: cloudflare
cf-ray: 8974a39fb8cdabc9-CPH
alt-svc: h3=":443"; ma=86400

curl: (22) The requested URL returned error: 502
Warning: Problem : HTTP error. Will retry in 512 seconds. 5 retries left.
curl: (28) Resolving timed out after 2000 milliseconds
Warning: Problem : timeout. Will retry in 600 seconds. 4 retries left.
curl: (28) Resolving timed out after 2000 milliseconds
Warning: Problem : timeout. Will retry in 600 seconds. 3 retries left.

![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/b7f35467-d711-40f1-a3d7-ee2adf38d2fc)
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/889297a1-29a8-4c3a-bd8a-8750a76f0b41)
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/00ad9f31-f24f-4b7a-a1c3-5a66ce36b60f)
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/3141ba33-3f46-483a-a282-a04e2159f3b4)


Praying that someone is able to help me solve this!
Thanks a lot in advance!

**Reply by gandhiarpit** [reply]
You have wrong domains in Let's encrypt config. I am not able to reach tesla public key, so that's something wrong. 

It should be 
```
 domains:
  - ha.apil.dk
  - "*.ha.apil.dk"
  - tesla.apil.dk
  - homeassistant-internal.apil.dk
```

**Reply by apildk** [reply]
Thank you a lot for your reply! <3

Though seems i still have an issue.

I have edited the let's config config as you said.


I deleted the two files from /ssl before starting up let's encrypt again


i have removed the "use_x_forwarded_for: true" from my configuration.yaml


then i have restarted the 3 addons, tesla http proxy, cloudflare and ngix ssl proxy

but i am still getting errors :(

![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/e8c59d7e-e057-45da-acef-b3c5bf755fcb)
![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/37192167/7421abe4-7cbf-40fa-89fa-5783dedea9a4)


Is there somewhere that i need to manually move some files to, or maybe i should delete everything and try from scratch again ?

Thanks again!

**Reply by gandhiarpit** [reply]
Oh sorry about HA. 
I was too scared to make it work on my Main HA due to proxy bringing down my proxmox host. So I span up a new HA instance on my unraid server exactly for this. That one only has couple of addons required for this proxy. 
And used [Remote HomeAssistant](https://github.com/custom-components/remote_homeassistant) component to bind it to my main HA.

If you have means to run another HA VM, you could try that.

**Reply by apildk** [reply]
No worries about the HA going down, did not take me long to fix that issue :)

I am considering starting over with this, and moving away from my rasp pi 4 to have HA on it with everything running as an addon, and then move to my synology nas to have things running on docker there.

**Comment by apildk** [comment]
Hi again!

I am still unable to make this work, hoping someone can help me out!

Best regards

**Comment by GerbenPapo** [comment]
I ended up adding a transform rule in Cloudflare.

Just like suggesting above earlier, generate the required certificate files using the Tesla HTTP Proxy, then copy /share/tesla/* to /homeassistant/www/tesla/

As this will produce URLs like https://example.com/local/tesla/com.tesla.3p.public-key.pem, use Cloudflare to rewrite that to the expected URL:
Rules -> Transform Rules -> Create rule:
1. Choose a name
2. Custom filter expression
3. When incoming requests match... URI Full - equals `https://tesla.example.com/.well-known/appspecific/com.tesla.3p.public-key.pem` (use your own domain)
4. Rewrite to... Static -> local/tesla/com.tesla.3p.public-key.pem
5. Query Preserve
6. Save.

Give it a couple of seconds to propagate and it should start serving the public key.

Disclaimer: I have not used this over a longer period of time yet.

**Reply by apildk** [reply]
Hi GerbenPapo.

Thanks a lot for your reply!

Can you see if i have done this correctly ?
![image](https://github.com/user-attachments/assets/ab72e158-221b-4dd1-8afe-a997a1917b34)

I am still getting the 502 error.
![image](https://github.com/user-attachments/assets/90e8f220-4d8a-4b64-86a8-4261e94c06f3)


Thanks again!
Best regards
Apil

**Reply by GerbenPapo** [reply]
The transform rule looks fine, but Cloudflare cannot reach tesla.apil.dk (your server/ha)
This can have many reasons. Some obvious ones:
Have you configured Cloudflared that it listens on tesla.apil.dk? 
Is the tunnel set up correctly, ie does it show its active/online in the Cloudflare panel?
Firewall perhaps?

**Reply by apildk** [reply]
Hmm i am not sure which parts to check.
But ha.apil.dk which links to my HA works fine with the tunnel.
![image](https://github.com/user-attachments/assets/d9b45783-07a8-4238-989c-e8dc028c71cd)
![image](https://github.com/user-attachments/assets/7c51015c-f496-41d4-9ffb-d860faecaf39)
![image](https://github.com/user-attachments/assets/cb2bd1a9-a9ff-4bc3-ab8e-eb11d967d79f)
![image](https://github.com/user-attachments/assets/5826a0e2-4a5d-4aff-9bd5-4b7da0d9f239)

I dont think my firewall in my router should interfeer here, as it does not with connecting to HA.

Thanks again!

**Reply by GerbenPapo** [reply]
That all looks just fine. And did you also define the hostname tesla.apil.dk in the tunnel config?

**Reply by apildk** [reply]
Hmm tunnel config ?
Is that the config of additional hosts in the cloudflare addon as here :
![image](https://github.com/user-attachments/assets/4fcaffae-b2de-4ddb-b359-d85b2a99888f)
The DNS config here:
![image](https://github.com/user-attachments/assets/fac5b9f1-bac9-4540-bb1f-fff97a0c8d65)
In the Tesla HTTP Proxy Addon:
![image](https://github.com/user-attachments/assets/e0ea5933-109e-45bc-b936-eb346fd91071)
Lets encrypt addon:
![image](https://github.com/user-attachments/assets/b014ac90-7613-4c5c-a4a2-9ea743e628d6)


Sounds like i am overlooking somewhere cruicial to define it, that i am unware off ?

**Reply by GerbenPapo** [reply]
There’s a section in the tunnel config called something like “public hostname”
It needs the subdomain there as well

**Reply by apildk** [reply]
Okay, hmm when i go to the zero trust dashboard and press configure, it wants to start a guide with migration for some reason, anyways, udring the step two it wants me to preview the origin configurations of each hostname, i see here that tesla.apil.dk is on the list of public hostnames.
![image](https://github.com/user-attachments/assets/ab0f479e-e9e8-4259-9a48-587af97e2abd)

I will try and finalize the migration

**Reply by apildk** [reply]
Finalized the migration, and yea, i think it looks correct ?
(Did not add the tesla.apil.dk, has been there "all the time")
![image](https://github.com/user-attachments/assets/865f846d-8f61-404a-8f5d-ca188f722715)

**Reply by GerbenPapo** [reply]
I think I see the problem. Tesla.apil.dk is pointing to the internal and non-proxied domain. This will work from inside your network, but not from the public internet. 
Try updating it to http://homeassistant:8123

**Reply by apildk** [reply]
So like this ?
![image](https://github.com/user-attachments/assets/0cffe675-d6d5-45e2-ba78-562d0e470d0d)

It should be correct though from what the guide says (https://github.com/llamafilm/tesla-http-proxy-addon/wiki/Cloudflare-Argo-Tunnel)
![image](https://github.com/user-attachments/assets/ada3f9a5-df9b-42fc-b9d9-76250ddc9e4d)

Hmm, problem seems to persist.
![image](https://github.com/user-attachments/assets/2faa10ac-635f-4b26-9c22-35c85e9fbfd7)
![image](https://github.com/user-attachments/assets/59f8783b-51f7-4784-97a7-08fd2d1edb98)
![image](https://github.com/user-attachments/assets/eebfabf4-0077-447c-a877-432a21bc3a13)

Here is nginx ssl proxy
![image](https://github.com/user-attachments/assets/c8abbea9-4e41-46a6-93f5-48680b827865)
![image](https://github.com/user-attachments/assets/42cb761d-f821-432e-b148-c9abdf4fc5cd)

**Reply by GerbenPapo** [reply]
No, the stuff in you ha config was ok. The Cloudflare tunnel config in the zero trust panel needs to change. 
also, I’m not sure if you need any of the nginx ssl proxy stuff (I don’t use it)

**Reply by apildk** [reply]
Im so sorry for my misunderstanding.
Like this then :) ? :

And http, not https ?

**Reply by apildk** [reply]
I think that did it! i now get the option to download the certificate! :-D

**Reply by ballonchef** [reply]
@apildk : How did you edit the cloudflare tunnel config? Where did you find that? I can't find it in my cloudflare account.

EDIT: I could use the cloud flare manager to change the tunnel settings. However, I'm not sure if I should migrate to a "remotely managed tunnel", as this is irreversible. How did you do it?

**Reply by steve-hawkins** [reply]
> No, the stuff in you ha config was ok. The Cloudflare tunnel config in the zero trust panel needs to change. also, I’m not sure if you need any of the nginx ssl proxy stuff (I don’t use it)

@GerbenPapo could you explain your set up a little further please?  When you say you don't use the nginx ssl proxy how have you got tesla http proxy running?

I have got as far as successfully setting up the tesla public hostname in the cloudflare tunnel config and that can reach my HA server, but when I follow the tesla http proxy set up docs by placing the tesla.mydomain.com only in the config and starting the addon it suggests after a few seconds it should stop and the certificate be available here `https://tesla.mydomain.com/.well-known/appspecific/com.tesla.3p.public-key.pem`

the addon doesn't stop itself and I get 404 when trying to reach here `https://tesla.mydomain.com/.well-known/appspecific/com.tesla.3p.public-key.pem` I tried restarting the tesla http proxy addon, but same results.

any help appreciated

**Reply by GerbenPapo** [reply]
The Tesla proxy only cares about validating the public key endpoint (your instance) during start up. If that ends up in a 404, the add-on will stop running. 
I’m not using the nginx ssl proxy, because afaik this is only needed to provide valid https endpoint for the public key. Cloudflare essentially is an ssl proxy already, hence I decided not to use two proxies. 

You’ll need to place the files in a folder where it’s publicly accessible. HA has a media folder that’s called something like “local” (check the earlier posts). If you place the key files there, and the adjust the Cloudflare route as described above. 

for me this has been working flawlessly for the last couple of weeks
As I’m currently not near a computer or the car for an extended period of time, I’m not able to make a better explanation or decorate this post with screenshots, so I hope the blurb above is making sense. 

But, there may be an even better solution yet: as of ha 2024.08 the Tesla fleet API is officially supported as an integration/service by HA. This looks to simplify the setup by a lot, but I haven’t been able to test drive this myself. Maybe you have better luck with this. Let me know how it goes (with either of the attempts)!

**Reply by steve-hawkins** [reply]
I just tried the new [HA Tesla Fleet integration](https://www.home-assistant.io/integrations/tesla_fleet/) and the set up couldn't be easier:

- install integration
- log into Tesla account
- accept all permissions scopes
- done!!!

just going through updating all my HA entities and will report back on any feature / stability differences

**Reply by GerbenPapo** [reply]
Thanks for trying it out and I’m happy to hear the initial results are promising! I’ll be switching to the official supported integration in a week or two. Good luck updating the rest of the entities!

**Reply by JeandreRoux** [reply]
I also just had a go with the Official Tesla Fleet Integration but was disappointed that it only reports sensor data. I can't send any commands to the car, which is what I want. This will probably be added later but for now I'll try your CloudFlare redirect method.

**Reply by JeandreRoux** [reply]
So I got most of it to work. In the CloudFlared add-on I had to put tesla.example.com directs to the local IP address of my HA instance so that when the CloudFlare rule kicks in the local path actually goes to the public key.
That all worked fine and I got approved by Tesla Development.
But now I'm stuck at setting up the virtual key through the Tesla HTTP Proxy Addon. When I click 'Enroll public key in your vehicle' it doesn't open the Tesla App, just takes me to a webpage telling me to download the Tesla App, sign in and then come back and click finish setup. Clicking Finish setup doesn't do anything.

Help would be greatly appreciated. I think you said you don't run the Tesla HTTP Proxy, if so, how do you complete the virtual key setup?
Thanks!

**Reply by JeandreRoux** [reply]
I figured it out, I opened the 'Enroll public key in your vehicle' with my laptop which then generated a QR code. I scanned it with my phone and it opened the Tesla App and setup the Virtual Key. All appears to be working now.

Thanks for the figuring out that you did to get this to work.

**Reply by JeandreRoux** [reply]
Just me again... did I do something wrong above?
I'm getting data from the car fine but I can't send any commands. Last night when I first setup the virtual key I was able to send commands to the car for a few minutes but now its not working... Any ideas?

**Reply by GerbenPapo** [reply]
Maybe you're being rate limited for sending too many commands. You can check the logs in the add-on if this is the case.

Just to be sure: Are you using the official integration (Tesla fleet), or the Proxy Add-on? If the first, you'd be better off taking the conversation there(https://github.com/home-assistant/core/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+tesla_fleet%22)
